File: selfie-camera/src/main/java/org/odk/collect/selfiecamera/CaptureSelfieActivity.java
/*
 * Copyright 2017 Nafundi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.odk.collect.selfiecamera;

import android.os.Bundle;
import android.view.Window;
import android.view.WindowManager;

import org.odk.collect.androidshared.ui.ToastUtils;
import org.odk.collect.permissions.PermissionsProvider;
import org.odk.collect.shared.injection.ObjectProvider;
import org.odk.collect.shared.injection.ObjectProviderHost;
import org.odk.collect.strings.localization.LocalizedActivity;

public class CaptureSelfieActivity extends LocalizedActivity {

    public static final String EXTRA_TMP_IMAGE_PATH = "tmpImagePath";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ObjectProvider objectProvider = ((ObjectProviderHost) getApplication()).getObjectProvider();
        PermissionsProvider permissionsProvider = objectProvider.provide(PermissionsProvider.class);

        if (!permissionsProvider.isCameraPermissionGranted()) {
            finish();
            return;
        }

        requestWindowFeature(Window.FEATURE_NO_TITLE);
        getWindow().setFlags(WindowManager
                .LayoutParams.FLAG_FULLSCREEN, WindowManager.LayoutParams.FLAG_FULLSCREEN);
        setContentView(R.layout.activity_capture_selfie);
        if (null == savedInstanceState) {
            Camera2Fragment fragment = new Camera2Fragment();
            Bundle args = new Bundle();
            args.putString(Camera2Fragment.ARG_TMP_IMAGE_PATH, getIntent().getStringExtra(EXTRA_TMP_IMAGE_PATH));
            fragment.setArguments(args);

            getFragmentManager().beginTransaction()
                    .replace(R.id.container, fragment)
                    .commit();
        }
        ToastUtils.showLongToast(this, R.string.take_picture_instruction);
    }
}

File: selfie-camera/src/main/java/org/odk/collect/selfiecamera/CaptureSelfieActivity.kt
/*
 * Copyright 2017 Nafundi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.odk.collect.selfiecamera

import android.Manifest
import android.os.Bundle
import android.view.View
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import org.odk.collect.analytics.Analytics
import org.odk.collect.androidshared.ui.ToastUtils.showLongToast
import org.odk.collect.externalapp.ExternalAppUtils
import org.odk.collect.permissions.PermissionsChecker
import org.odk.collect.strings.localization.LocalizedActivity
import javax.inject.Inject

class CaptureSelfieActivity : LocalizedActivity() {

    @Inject
    internal lateinit var stillCamera: StillCamera

    @Inject
    internal lateinit var videoCamera: VideoCamera

    @Inject
    lateinit var permissionsChecker: PermissionsChecker

    override fun onCreate(savedInstanceState: Bundle?) {
        (application as SelfieCameraDependencyComponentProvider)
            .selfieCameraDependencyComponent
            .inject(this)

        super.onCreate(savedInstanceState)

        if (!permissionsGranted()) {
            finish()
            return
        }

        val insetsController = WindowCompat.getInsetsController(window, window.decorView)
        insetsController.hide(WindowInsetsCompat.Type.statusBars())
        setContentView(R.layout.activity_capture_selfie)

        val previewView = findViewById<View>(R.id.preview)

        val camera = if (intent.getBooleanExtra(EXTRA_VIDEO, false)) {
            videoCamera
        } else {
            stillCamera
        }

        camera.initialize(this, previewView)
        camera.state().observe(this) {
            when (it) {
                Camera.State.UNINITIALIZED -> {}
                Camera.State.INITIALIZED -> setupCamera(camera)
                Camera.State.FAILED_TO_INITIALIZE -> {
                    showLongToast(this, R.string.camera_failed_to_initialize)
                }
            }
        }
    }

    private fun setupCamera(camera: Camera) {
        val previewView = findViewById<View>(R.id.preview)

        if (camera is VideoCamera) {
            val videoPath = intent.getStringExtra(EXTRA_TMP_PATH) + "/tmp.mp4"
            previewView.setOnClickListener {
                if (!camera.isRecording()) {
                    camera.startVideo(
                        videoPath,
                        { ExternalAppUtils.returnSingleValue(this, videoPath) },
                        { showLongToast(this, R.string.camera_error) }
                    )

                    showLongToast(this, getString(R.string.stop_video_capture_instruction))
                } else {
                    camera.stopVideo()
                }
            }

            showLongToast(this, getString(R.string.start_video_capture_instruction))
            Analytics.log(AnalyticsEvents.RECORD_SELFIE_VIDEO, "form")
        } else if (camera is StillCamera) {
            val imagePath = intent.getStringExtra(EXTRA_TMP_PATH) + "/tmp.jpg"
            previewView.setOnClickListener {
                camera.takePicture(
                    imagePath,
                    { ExternalAppUtils.returnSingleValue(this, imagePath) },
                    { showLongToast(this, R.string.camera_error) }
                )
            }

            showLongToast(this, R.string.take_picture_instruction)
        }
    }

    private fun permissionsGranted(): Boolean {
        return if (intent.getBooleanExtra(EXTRA_VIDEO, false)) {
            return permissionsChecker.isPermissionGranted(
                Manifest.permission.CAMERA,
                Manifest.permission.RECORD_AUDIO
            )
        } else {
            permissionsChecker.isPermissionGranted(Manifest.permission.CAMERA)
        }
    }

    companion object {
        const val EXTRA_TMP_PATH = "tmpPath"
        const val EXTRA_VIDEO = "video"
    }
}
